File: jupyter2/ps/memory_management/vis_addr/go/vis_addr/vis_addr.go
package main
import (
    "os"
    "strconv"
    "fmt"
    "github.com/loov/hrtime"
)

func get_int64(args []string, i int, def_val int64) int64 {
	if i < len(args)  {
		x, _ := strconv.Atoi(args[i])
		return int64(x)
	} else {
		return def_val
	}
}

func time_ns() int64 {
	return int64(hrtime.Now())
}

func main() {
	s := get_int64(os.Args, 1, int64(1000 * 1000))
	m := get_int64(os.Args, 2, int64(10))
	n := get_int64(os.Args, 3, m * 10)
	sizeof_int64 := int64(8)
	if sizeof_int64 * s * m > (1 << 30) {
		fmt.Fprintf(os.Stderr, "you'd better not allocate that much memory\n")
		fmt.Fprintf(os.Stderr, "sizeof(int64)(8) * s(%d) * m(%d) = %d > 1GB\n", s, m, sizeof_int64 * s * m)
		os.Exit(1)
	}
	a := make([][]int64, m)
	for i := int64(0); i < n; i++ {
		t0 := time_ns()
		b := make([]int64, s)
		t1 := time_ns()
		for j := range b { b[j] = i }
		a[i % m] = b
		fmt.Printf("%d %d %d\n", i, &b[0], t1 - t0)
	}
}

File: jupyter2/ps/memory_management/vis_addr/cc/vis_addr/vis_addr.cc
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

long time_ns() {
  struct timespec ts[1];
  clock_gettime(CLOCK_REALTIME, ts);
  return ts->tv_sec * 1000000000L + ts->tv_nsec;
}

int main(int argc, char ** argv) {
  long s = (1 < argc ? atol(argv[1]) : 1000 * 1000);
  long m = (2 < argc ? atol(argv[2]) : 10);
  long n = (3 < argc ? atol(argv[3]) : m * 10);
  if (sizeof(long) * s * n > (1L << 32)) {
      fprintf(stderr, "you'd better not allocate that much memory\n");
      fprintf(stderr, "sizeof(long)(8) * s(%ld) * n(%ld) < 4GB\n", s, n);
      exit(1);
  }
  long ** a = (long **)malloc(sizeof(long *) * m);
  for (long i = 0; i < n; i++) {
    long t0 = time_ns();
    long * b = (long *)malloc(sizeof(long) * s);
    long t1 = time_ns();
    for (long j = 0; j < s; j++) b[j] = i;
    a[i % m] = b;
    printf("%ld %ld %ld\n", i, (long)&b[0], t1 - t0);
  }
  return 0;
}

